Feature/configurable access message logging

aobermair opened the pull request.

dcseifert commented on internal/pkg/api/rest-handler.go:
```diff
@@ -63,8 +63,6 @@ func (proxy *restProxy) handleV1DataGet(w http.ResponseWriter, r *http.Request)
 func (proxy *restProxy) handleV1DataPost(w http.ResponseWriter, r *http.Request) {
 	// Add unique identifier for logging purpose
 	r = utilInt.AssignRequestUID(r)
```
I think we should either remove the request uid entirely or keep is somehow.
Because now we have the problem that the uid is logged inside the default-compiler if i.e. the request has a wrong format, but with not logging the initial request the uid becomes type of useless ;D

I think the best approach is to remove the entire uid thing from the incoming request header and log important fields like request URI including request body-fields like path or method inside the policy compiler itself.

dcseifert reviewed (changes requested):
All in all the changes look quite good, please have a look at my detailed code-review :D

aobermair commented on internal/pkg/opa/default-compiler.go:
```diff
@@ -145,11 +151,13 @@ func (compiler policyCompiler) ServeHTTP(w http.ResponseWriter, req *http.Reques
 	// OPA decided denied
 	if queries.Queries == nil {
 		compiler.writeDeny(w, uid)
+		logging.LogAccessDecision(compiler.accessDecisionLogLevel, req.RequestURI, req.Method, time.Since(startTime).String(), "DENY")
```
So the logging should contain the input path and method of the body?
I decided against moving into writeAllow and wirteDeny, cause I thought it is not a useful idea to add function params only for logging case. But i can move it into writeAllow and writeDeny, if you are ok with this.

dcseifert replied:
Good point. Or we just remove the uid and log the request body in case of an error.

aobermair commented on internal/pkg/opa/default-compiler.go:
```diff
@@ -26,10 +27,11 @@ import (
+	appConfig              *configs.AppConfig
+	config                 *opa.PolicyCompilerConfig
+	engine                 *OPA
+	accessDecisionLogLevel string
```
👍

aobermair commented on pkg/constants/logging/logging.go:
```diff
@@ -10,6 +10,43 @@ const LabelComponent string = "component"
+	return
+}
+
+func checkAccessDecisionLogLevel(logLevel, decision string) bool {
+	loggingStatus := false
+	switch logLevel {
+	case "ALL":
+		loggingStatus = true
+	default:
+		if logLevel == decision {
+			loggingStatus = true
+		}
+	}
+	return loggingStatus
+}
```
👍  Did not see this simplification after reducing the cases

aobermair commented:
- Added review improvements
- Removed uids (except uid helper, but no usage anymore?)  
- Changed path and method to body input path and method
- Added new logging to envoy-proxy

dcseifert commented on internal/pkg/api/envoy/envoy-proxy.go:
```diff
@@ -207,21 +210,16 @@ func (p *envoyExtAuthzGrpcServer) Check(ctx context.Context, req *ext_authz.Chec
-	httpRequest = utilInt.AssignRequestUID(httpRequest)
-	uid := utilInt.GetRequestUID(httpRequest)
-	logging.LogForComponentAndUID("envoyExtAuthzGrpcServer", uid).Infof("Received Envoy-Ext-Auth-Check to URL: %s", httpRequest.RequestURI)
-
 	w := telemetry.NewInMemResponseWriter()
 	(*p.compiler).ServeHTTP(w, httpRequest)
 
 	resp := &ext_authz.CheckResponse{}
 	switch w.StatusCode() {
 	case http.StatusOK:
-		logging.LogForComponentAndUID("envoyExtAuthzGrpcServer", uid).Infoln("Decision: ALLOW")
+		logging.LogAccessDecision(p.cfg.AccessDecisionLogLevel, path, r.GetMethod(), time.Since(startTime).String(), "ALLOW", "envoyProxy")
 		resp.Status = &rpc_status.Status{Code: int32(code.Code_OK)}
 	case http.StatusForbidden:
-		logging.LogForComponentAndUID("envoyExtAuthzGrpcServer", uid).Infoln("Decision: DENY")
+		logging.LogAccessDecision(p.cfg.AccessDecisionLogLevel, path, r.GetMethod(), time.Since(startTime).String(), "DENY", "envoyProxy")
```
As far as I see it right now, the envoy-proxy and the policy-compiler are both logging the access decision. Therefore I thing removing the logs from the envoy-proxy is the best option to avoid duplications (the rest-proxy doesn't log the decisions either).

Removing the Envoy-Proxy from Kelon will also be an issue for the next year (as I discussed with Matthias right now) so you shouldn't put too much effort in this ;D.

dcseifert commented on pkg/telemetry/applicationinsights.go:
```diff
@@ -169,7 +167,7 @@ func (p *ApplicationInsights) GetHTTPMiddleware() (func(handler http.Handler) ht
 			parentID := request.Header.Get("request-id")
 			trace.Tags.Operation().SetParentId(parentID)
-			trace.Tags.Operation().SetId(parentID + uid)
+			trace.Tags.Operation().SetId(parentID)
```
Removing the uid from the parentID will break the correlation logic. But due to the fact that we want to remove application insights anyway I would suggest to disable application insights for now and remove application insights in a later issue.

dcseifert commented on internal/pkg/opa/default-compiler.go:
```diff
@@ -93,84 +92,94 @@ func (compiler *policyCompiler) Configure(appConf *configs.AppConfig, compConf *
+		compiler.writeDeny(w)
+		logging.LogAccessDecision(compiler.config.AccessDecisionLogLevel, path.String(), method, time.Since(startTime).String(), "DENY", "policyCompiler")
 		return
 	}
 	// Check if any query succeeded
 	if done := anyQuerySucceeded(queries); done {
-		compiler.writeAllow(w, uid)
+		compiler.writeAllow(w)
+		logging.LogAccessDecision(compiler.config.AccessDecisionLogLevel, path.String(), method, time.Since(startTime).String(), "ALLOW", "policyCompiler")
 		return
 	}
 
 	// Otherwise translate ast
 	result, err := (*compiler.config.Translator).Process(queries, output.Datastore, req)
 	if err != nil {
-		compiler.handleError(w, uid, errors.Wrap(err, "PolicyCompiler: Error during ast translation"))
+		compiler.handleError(w, errors.Wrap(err, "PolicyCompiler: Error during ast translation"))
 		return
 	}
 
 	// If we receive something from the datastore, the query was successful
 	if result {
-		compiler.writeAllow(w, uid)
+		compiler.writeAllow(w)
+		logging.LogAccessDecision(compiler.config.AccessDecisionLogLevel, path.String(), method, time.Since(startTime).String(), "ALLOW", "policyCompiler")
 	} else {
-		compiler.writeDeny(w, uid)
+		compiler.writeDeny(w)
+		logging.LogAccessDecision(compiler.config.AccessDecisionLogLevel, path.String(), method, time.Since(startTime).String(), "DENY", "policyCompiler")
```
I think we should really think again about moving the logging logic into the writeDeny/Allow methods. Because changing three lines when changing something in the logic is a bit messy. You could encapsulate the arguments that are needed for logging in a separate struct 'decisionContext' or so to keep the API of these methods cleaner.

dcseifert commented on internal/pkg/opa/default-compiler.go:
```diff
@@ -93,84 +92,94 @@ func (compiler *policyCompiler) Configure(appConf *configs.AppConfig, compConf *
+	}
+
+	path, errPath := extractURLFromRequestBody(input)
+	method, errMethod := input["method"].(string)
```
I would be careful with just accessing a map and casting the result into the type string. I would implement some additional method like extractURLFromRequestBody() which returns some useful error messages in case of a missing but required field (Because this field is passed by the user of our service it could be malformed in many interesting ways we should definitely be prepared for :D)

dcseifert commented on cmd/kelon/kelon.go:
```diff
@@ -50,8 +50,9 @@ var (
-	logFormat = app.Flag("log-format", "Log-Format for Kelon. Must be one of [TEXT, JSON]").Default("TEXT").Envar("LOG_FORMAT").Enum("TEXT", "JSON")
+	logLevel               = app.Flag("log-level", "Log-Level for Kelon. Must be one of [DEBUG, INFO, WARN, ERROR]").Default("INFO").Envar("LOG_LEVEL").Enum("DEBUG", "INFO", "WARN", "ERROR", "debug", "info", "warn", "error")
+	logFormat              = app.Flag("log-format", "Log-Format for Kelon. Must be one of [TEXT, JSON]").Default("TEXT").Envar("LOG_FORMAT").Enum("TEXT", "JSON")
+	accessDecisionLogLevel = app.Flag("access-decision-log-level", "Access decision Log-Level for Kelon. Must be one of [ALL, ALLOW, DENY, NONE]").Default("ALL").Envar("ACCESS-DECISION-LOG-LEVEL").Enum("ALL", "ALLOW", "DENY", "NONE", "all", "allow", "deny", "none")
```
I would change the environment variable 'ACCESS-DECISION-LOG-LEVEL' to 'ACCESS_DECISION_LOG_LEVEL' to preserve a uniform variable style.

dcseifert reviewed (changes requested):
Well done. I just have some minor changes before we can finish this PR :).

dcseifert reviewed (approved):
Looks great!
